### DISCUSSION TITLE
Correct placement of user logs generation

### DISCUSSION THREAD

**Discussion by JaconsMorais** [adonisjs/core#2696] [Help] [answered]
Hi all, I have been implementing some user logs in my app (like 'user created product A' or 'user A approved login of user B'), but I did not figure out where is the correct place to put my instruction to create and save log in database.

I have two models, User and UserLog

The relation of this two models is:
User hasMany UserLogs
UserLogs belongsTo User.

At first, I tried to put inside authorize method in custom validator, but I realize that the validator runs before controller, and the log could be created even the request returns any errors.

So I tried to put in Controller, but I dont know if is the right place. Maybe some file that runs after controller in request life cycle, but I dont know.

I tried to put inside some hook, but I need some infos like request.ip() or auth.user, and its not available on hooks.

I know adonis has built in support for logging, but all things I founded in docs was about system logs, not user logs.

What is the correct approach to achieve this?

**Comment by d4n1b** [comment, marked as answer]
Hi @JaconsMorais ,
It sounds like the perfect use for [events](https://docs.adonisjs.com/guides/events). You can call it from anywhere you have access to  [request ctx](https://docs.adonisjs.com/guides/context).

I'm thinking of something around:

ProductsController.ts
```
import Event from '@ioc:Adonis/Core/Event';

export default class ProductsController {
  public store(ctx) {
    // save logic...

    Event.emit('user:event', {
      message: `user created ${product.name}`,
      ctx,
    });
  }
}
```

start/event.ts
```
Event.on('user:event', ({ message, ctx }) => {
  // Call UserLogs model here
  console.log(`${message} from IP ${ctx.request.ip()}`);
});
```

I hope it helps.
